texts/0005-tauri-config-restructure.md

```diff
@@ -0,0 +1,170 @@
+- Move `tauri > cli` and `tauri > update` fields to `plugins > cli` and `plugins > update` as they are plugins now.
+- Rename `build > distDir` to `frontendDist` to explicitly set the intent of the option.
+- Rename `build > devPath` to `build > devUrl` and only accept urls,
+  if users don't have a devServer, they should removve this field and only set `build > dist` which will make the CLI
+  start its built-in devServer or fallback to embed the assets if `--no-dev-server` is used.
+
+<table>
+<thead>
+  <tr>
+    <th>Current Config</th>
+    <th>New Config</th>
+  </tr>
+</thead>
+<tbody>
+  <tr>
+  <td>
+
+```jsonc
+{
+  "build": {
+    "beforeDevCommand": "pnpm dev",
+    "beforeBuildCommand": "pnpm build",
+    "devPath": "http://localhost:8080/",
+    "distDir": "../src",
+    "withGlobalTauri": true
+  },
+  "package": {
+    "productName": "tauri-app",
+    "version": "0.0.0"
+  },
+  "tauri": {
+    "allowlist": {
+      "all": true
+    },
+    "bundle": {
+      "active": true,
+      "targets": "all",
+      "identifier": "com.taasduri.dev",
+      "icon": [
+        "icons/32x32.png",
+        "icons/128x128.png",
+        "icons/128x128@2x.png",
+        "icons/icon.icns",
+        "icons/icon.ico"
+      ]
+    },
+    "updater": {
+      "active": true,
+      "pubkey": "",
+      "endpoints": ["http://localhost:8080/update.json"],
+      "windows": {
+        "installMode": "basicUi"
+      }
+    },
+    "cli": {},
+    "macOSPrivateApi": false,
+    "security": {
+      "csp": null
+    },
+    "pattern": {
+      "use": "brownfield"
+    },
+    "systemTray": {
+      "iconPath": "./path/to/icon"
+    },
+    "windows": [
+      {
+        "title": "tauri-app",
+        "width": 800,
+        "height": 600
+      }
+    ]
+  }
+}
+```
+
+</td>
+<td>
+
+```jsonc
+{
+  "productName": "tauri-app",
+  "version": "0.1.0",
+  "identifier": "com.tauri.dev",
+  "build": {
+    "beforeDevCommand": "pnpm dev",
+    "beforeBuildCommand": "pnpm build",
+    "devUrl": "http://localhost:8080/",
+    "frontendDist": "../src" // or ["../src/index.html", "../src/main.js"] if need to include specific files
+  },
+  "withGlobalTauri": true,
+  "macosPrivateApi": false,
+  "windows": [
+    {
+      "title": "tauri-app",
+      "width": 800,
+      "height": 600
+    }
+  ],
+  "trayIcon": {
+    "iconPath": "./path/to/icon"
+  },
+  "allowlist": { // soon to be replaced with permissions and cababilities
+    "all": true
+  },
+  "security": {
+    "pattern": "brownfield",
+    "csp": null
+  },
+  "plugins": {
+    "cli": {},
+    "updater": {
+      "pubkey": "",
+      "endpoints": ["http://localhost:8080/update.json"],
+      "windows": {
+        "installMode": "basicUi"
+      }
+    }
+  },
+  "bundle": {
+    "active": true,
+    "targets": "all",
+    "icon": [
+      "icons/32x32.png",
+      "icons/128x128.png",
+      "icons/128x128@2x.png",
+      "icons/icon.icns",
+      "icons/icon.ico"
+    ]
+  }
+}
+```
```

Comment:
```suggestion
```jsonc
{
  "metadata": {
    "identifier": "com.tauri.dev",
    "productName": "tauri-app",
    "version": "0.1.0",
  },
  "app": {
    "windows": [
      {
        "title": "tauri-app",
        "width": 800,
        "height": 600
      }
    ],
    // soon to be replaced with permissions and capabilities
    "allowlist": {
      "all": true
    },
    "security": {
      "pattern": "brownfield",
      "csp": null
    },
    "trayIcon": {
      "iconPath": "./path/to/icon"
    },
    "withGlobalTauri": true,
    "macosPrivateApi": false
  },
  "delivery": {
    "build": {
      "beforeDevCommand": "pnpm dev",
      "beforeBuildCommand": "pnpm build",
      "devUrl": "http://localhost:8080/",
      // or ["../src/index.html", "../src/main.js"] if need to include specific files
      "frontendDist": "../src"
    },
    "bundle": {
      "active": true,
      "targets": "all",
      "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ]
    }
  },
  "plugins": {
    "cli": {},
    "updater": {
      "pubkey": "",
      "endpoints": ["http://localhost:8080/update.json"],
      "windows": {
        "installMode": "basicUi"
      }
    }
  },
}
```
```
In my opinion, I more like to separate fields by "where/when would use it(mainly)" 
which I think is fits more in the app developer's mental model

like separate into four big group: 

- metadata: information that doesn't control/effects app logic
  (most of the time, or at least, not intend to)
- app: controls/effects how the app behaves
- delivery: control/effects how the app is build and bundled
- plugins: plugins configs

a side note: maybe extract the `beforeDevCommand` and `devUrl` 
  to the "develop" section before "delivery" section, 
  for group things that only will use for/effect dev process, 
  like a field for .ignore for `tauri dev` command to let devs to find it easier

Reply:
The grouping sounds reasonable IMO, except `plugins` should be under `app` and `metadata` should still be unpacked to the root, no need to have it in a `metadata`?

> a side note: maybe extract the `beforeDevCommand` and `devUrl` 
  to the "develop" section before "delivery" section, 
  for group things that only will use for/effect dev process, 
  like a field for .ignore for `tauri dev` command to let devs to find it easier
  
last time I checked, this was not desirable and keeping `beforeDevCommand` and similar close to each other is more desired.

Reply:
I am not sure I quite like `delivery` tbh, but let me mention some people to see if this desired. 

cc @lucasfernog @FabianLars @simonhyll

Reply:
In my opinion, a plugin may do things that across multiple section, 
like the updater is both in app logic and the bundle logic, 
so what I think is separate things in plugin config it self, if necessary

group into `metadata` just for avoid mixing with other thing, afaik json doesn't enforce field order

Reply:
If we keep just those 3 things in metadata i agree with Amr that it makes sense to leave that at is, if we split up the identifier into 3 seperate components (we talked about that ages ago) i think grouping them makes sense.

A bit torn about the plugins though, i think both (top level and part of app) could work but to keep the nesting at a saner level i think i **slightly** prefer top level. But yeah, i can totally see both work here.

Reply:
> Wasn't a big part of this rfc to get rid of the big groupings at the top levels? I am not totally against it, just unexpected i guess.

My understand of the rfc description let me thinks of some simple rule for let config items easy to find at intuition level, and clear-er separation to parts that not interact very much

I think follow a simple tree is easier than find in a relative big list

Reply:
> Wasn't a big part of this rfc to get rid of the big groupings at the top levels? I am not totally against it, just unexpected i guess.

Not really, the main reason is to make sense of why an option is there and keeping it consistent so it is easier to add more features later on.

Reply:
how about this then:
```jsonc
{
  "productName": "Tauri App",
  "identifier": "com.tauri.app", // and later we can make this take an object instead of string if needed
  "version": "0.0.0",
  "build": {
    "beforeDevCommand": "pnpm dev",
    "beforeBuildCommand": "pnpm build",
    "devUrl": "http://localhost:8080/",,
    "frontendDist": "../src" // or ["../src/index.html", "../src/main.js"] if need to include specific files
  },
  "app": {
    "withGlobalTauri": true,
    "windows": {},
    "plugins": {},
  },
  "bundle": {}
}
```

Reply:
i think i'd like that, though i'm not quite sold on the identifier thing. Making it accept an object in the future feels a bit improvised because the 3 fields are a more general property than just an identifier if that makes sense. This is the issue i was talking about btw https://github.com/tauri-apps/tauri/issues/6062 but i'd ignore the corrected paths it talks about for now since that would require some kind of migration to prevent dataloss.

idk maybe it's not that bad 🤔 let's see what the others think

Reply:
PR updated with the structure mentioned above